[2단계 - 장바구니 협업] 소피아(전소이) 미션 제출합니다.

soyi47 opened the pull request:
안녕하세요~ 리뷰어님, 소피아입니다. 😊

step2로 병민과 함께 페어로 진행해봤습니다.
일주일 넘게 페어를 진행하는 게 체력적으로는 너무 힘들었는데요. 그래도 혼자였다면 해보지 않았을 여러 시도나 생각을 해볼 수 있어서 좋았습니다.

- 배포 못 했습니다! 로컬에서 확인 부탁드립니다!
- 주말 쯤엔 연결해둔 백엔드 서버가 닫힌다고 들었습니다. 🤣 msw로 연결해서 올립니다! 
- [유저 API 명세](https://winter-rumba-a4b.notion.site/API-22194a5d7fc74cf7b0ebb5b4acebce62)
- [장바구니 API 명세](https://winter-rumba-a4b.notion.site/API-e84292d1f738480bbb7d901f03a51ecd)

<br >

### 🤔 고민한 부분들
일관성을 위해 모든 비동기 처리를 리덕스를 거치도록 수정해봤습니다. 이전 미션에서도 느꼈지만, 지금과 같이 상품 목록, 장바구니 아이템 목록 등을 굳이 전역 상태로 관리해야하는가에 대해 의문이 많이 생기고, 또 오히려 불편하다고 많이 느꼈습니다. 전역 상태로 관리하니 요청 전후로 전역 상태를 초기화해주거나, 너무 많은 상태를 관리해야하는 등의 불편함이 점점 더 늘어나네요. useFetch를 작성하고 가능한 페이지 단위로 비동기 처리하는 방향이 더 깔끔하다고 느껴져서, user 정보 외에는 리덕스에서 걷어내는 방향으로 수정해보고자 합니다.

pocojang committed:
feat: add contents

pocojang committed:
fix typos

pocojang committed:
feat: remove requirements

airman5573 committed:
chore: git 커밋 메세지의 규칙을 설정한다

airman5573 committed:
chore: 프로젝트 기본 설정

1. webpack
2. babel
3. scss
4.emotion
5. eslint & prettier

airman5573 committed:
feat(component): header를 생성한다

airman5573 committed:
feat(ui): home page 생성

airman5573 committed:
feat(cart): 장바구니에 상품 저장하는 기능 추가

airman5573 committed:
feat(redux): redux로 장바구니 상태 관리를 한다

airman5573 committed:
style(responsive): 반응형 디자인을 적용한다

airman5573 committed:
style(component): header를 sticky하게 만든다

airman5573 committed:
fix: link 수정

airman5573 committed:
refactor(style): 불필요한 style삭제

airman5573 committed:
docs(requirements): 기능 요구사항 목록을 생성한다

airman5573 committed:
chore(gitignore): tracking 하지 않을 목록을 추가한다

airman5573 committed:
refactor(redux): action type을 그룹핑한다

airman5573 committed:
refactor(a11y): image에alt속성을 추가한다

airman5573 committed:
refactor(style): css-in-js대신 css-module을 사용한다

airman5573 committed:
feat: live server대신 msw mocking server를 이용한다

airman5573 committed:
feat(ui): 이미지를 로딩하는 동안 indicator를 보여준다

airman5573 committed:
chore(webpack): 웹팩 설정 파일을 분리한다

airman5573 committed:
feat(style): windicss를 적용한다

airman5573 committed:
refactor(environment): 환경 변수를 수정한다

airman5573 committed:
refactor(cleanup): 불필요한 파일을 삭제한다

airman5573 committed:
style(favicon): favicon을 업데이트 한다

airman5573 committed:
refactor(cleanup): 불필요한 파일을 삭제한다

airman5573 committed:
refactor: className을 일관성 있게 작성한다

airman5573 committed:
refactor(cleanup): 절대 경로를 사용한다

airman5573 committed:
refactor(cleanup): 불필요한 className을 삭제한다

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
fix(title): 타이틀을 변경한다

youngbeomrhee committed:
Merge pull request #97 from airman5573/airman5573

[1단계 - 장바구니] 병민(윤병인) 미션 제출합니다.

airman5573 committed:
feat(storybook): storybook을 추가한다

airman5573 committed:
refactor(style): windicss 대신 tailwind를 사용한다

airman5573 committed:
refactor(cleanup): 불필요한 코드를 삭제한다

airman5573 committed:
refactor(component): dynamic한 css속성을 부여한다

airman5573 committed:
style(component): 왼쪽 여백을 추가한다

airman5573 committed:
refactor(component): handler함수를 등록한다

airman5573 committed:
refactor(lint): lint 수정

airman5573 committed:
refactor(cleanup): 불필요한 파일을 삭제한다

airman5573 committed:
fix(component): 함수가 있는 경우에만 호출한다

airman5573 committed:
refactor(component): icon에 button을 입힌다

airman5573 committed:
style(component): 기본 크기를 지정한다

airman5573 committed:
style(component): 스타일을 수정한다

airman5573 committed:
refactor(component): input에 variation을 추가한다

airman5573 committed:
feat(component): cart item을 추가한다

airman5573 committed:
refactor(lint): lint를 수정한다

airman5573 committed:
refactor(component): 속성값을 수정한다

airman5573 committed:
chore(storybook): 버전 업데이트를 한다

airman5573 committed:
chore(tailwind): pixel값을 사용한다

tailwind에서 기본으로 제공하는 util대신 pixel단위의 util class를 사용한다

airman5573 committed:
feat(storybook): tailwind를 스토리북에서도 사용한다

airman5573 committed:
style(component): css를 개선한다

airman5573 committed:
refactor(component): default props를 부여한다

airman5573 committed:
style(component): minor change

airman5573 committed:
refactor(component): 외부로부터 className을 받는다

airman5573 committed:
refactor(cleanup): 사용하지 않는 코드를 삭제한다

modal을 삭제한다. 지금당장 사용할 컴포넌트가 아니기 때문이다

airman5573 committed:
refactor(cleanup): 사용하지 않는 코드를 삭제한다

modal을 삭제한다

airman5573 committed:
refactor(component): 일관된 값을 넣어준다

airman5573 committed:
refactor(redux): 초기값을 변경한다

airman5573 committed:
refactor(webpack): 절대 경로를 추가한다

airman5573 committed:
refactor(component): props를 추가한다

체크되었는지 상태를 받는다

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
feat(component): 장바구니 form을 추가한다

airman5573 committed:
fix(webpack): 절대 경로를 추가한다

storyboard 뿐만 아니라 webpack에도 추가를 해준다

airman5573 committed:
refactor(typo): minor change

airman5573 committed:
refactor(cleanup): 사용하지 않는 prop을 삭제한다

airman5573 committed:
refactor(naming): 짧은 이름으로 변경한다

이름에 중복이 있기 때문이다

airman5573 committed:
refactor(api): api 구조를 변경한다

1.  title -> name : 상품의 "이름" 이 더 적절해보인다
2. price 단위 제거 : 더 유연하게 사용 가능하고 계산시에 편리하다
3. id -> sku : 상품을
식별하는 요소는 sku이다 id는 DB에서 사용하는 용어이다

airman5573 committed:
chore: 절대 경로를 추가한다

airman5573 committed:
style(component): 이미지의 최대 너비를 제한한다

airman5573 committed:
style(component): highlighter

스타일을 변경한다

ISSUES CLOSED: 기본 색상과 투명도를 수정했다

airman5573 committed:
refactor(utils): 유틸 함수로 분리한다

CartTotal컴포넌트에서 사용될것이기 때문에 분리한다

airman5573 committed:
style(component): 색상을 수정한다

airman5573 committed:
feat(component): 컴포넌트를 추가한다

총 주문 금액과 주문버튼이 있는 CartTotal컴포넌트를 추가한다

airman5573 committed:
fix(component): module을 가져오는 출처를 변경한다

airman5573 committed:
refactor(utils): 유틸 함수로 분리한다

airman5573 committed:
feat(component): 컴포넌트를 추가한다

alert 컴포넌트를 추가했다

airman5573 committed:
refactor(lint): lint를 바탕으로 수정한다

airman5573 committed:
fix(component): 올바른 컴포넌트를 사용한다

anchor대신 react-dom-router의 Link를 사용한다

airman5573 committed:
refactor(component): 상태 관리를 상위에 맡긴다

airman5573 committed:
refactor(component): 상태값이 있는 경우에 렌더링 한다

airman5573 committed:
refactor(component): 컴포넌트를 분리한다

airman5573 committed:
fix: 올바른 속성을 사용한다

airman5573 committed:
fix: 올바른 값을 입력한다

airman5573 committed:
fix(ui): 우선순위를 부여한다

이 부분은 추후에 다시 살펴봐야 한다. 이상하게 일반 태그의 우선순위가 더 높게 책정된다.

airman5573 committed:
chore(tailwind): 추가 css를 설정한다

airman5573 committed:
refactor(component): 외부에서 스타일링을 할 수 있도록 열어놓는다

airman5573 committed:
fix(redux): 장바구니에서 상품 삭제시 localStorage에도 반영한다

airman5573 committed:
refactor(redux): 상위에서 state를 가져온다

주소창에서 바로 장바구니로 넘어간 경우 ProductList를 거쳐가지 않았기 때문에 상품 정보가 없다. 그래서 상위에서 상품 정보를 가져온다

airman5573 committed:
perf(redux): re-render를 막는다

사실 자신이 없다. 정말 re-render를 막아줄까

airman5573 committed:
refactor(cleanup): 사용하지 않는 메뉴를 삭제한다

airman5573 committed:
test(redux): reducer 테스트를 추가한다

airman5573 committed:
Merge branch 'airman5573' of https://github.com/airman5573/react-shopping-cart into airman5573
Merge

airman5573 committed:
docs: 기능 요구 사항을 추가한다

airman5573 committed:
style(responsive): 헤더를 반응형으로 만든다

airman5573 committed:
style(component): 높이를 조정한다

flexbox를 쓰면 쫘악 위아래로 붙는데 이거를 해결한다

airman5573 committed:
style(component): 스크롤이 길어졌을때도 총금액 박스는 붙어있도록 한다

airman5573 committed:
refactor(style): style을 scss에 분리한다

airman5573 committed:
style(responsive): cart page를 반응형으로 개선한다

airman5573 committed:
style: 기본 좌우 padding을 준다

airman5573 committed:
refactor(component): 단독으로 사용도 가능하고 외부와 연결도 되게끔 개선한다

항상 외부에만 연결되어 사용하는 것이 아니라, 단독으로 사용해도 되게끔 한다. 직접 element에 접근해서 값을 가져다가 쓸 수도 있기 때문이다.

airman5573 committed:
chore(storybook): 배포 process를 추가한다

airman5573 committed:
refactor: es6 방식으로 import한다

airman5573 committed:
chore(git): tracking하지 않을 파일을 추가한다

airman5573 committed:
refactor(cleanup): 사용하지 않는 컴포넌트를 삭제한다

airman5573 committed:
refactor(readability): hook을 코드 상단으로 올린다

airman5573 committed:
refactor(component): 컴포넌트를 합친다

가독성을 위해 컴포넌트를 나누지 않고 합친다

airman5573 committed:
style(ui): 이미지 비율을 가로길이에 맞춘다

airman5573 committed:
style(ui): 장바구니 아이콘 배열을 조정한다

airman5573 committed:
fix(component): props를 올바르게 정의한다

airman5573 committed:
perf(component): 불필요한 동일성 체크를 제거한다

airman5573 committed:
perf(redux): 동일성 체크 함수를 개선한다

airman5573 committed:
chore(lint): 설정을 변경한다

dot-notation보다 [key]가 더 자연스러운 경우가 있기 때문에, off시켜준다

airman5573 committed:
test(reducer): test case를 추가한다

airman5573 committed:
chore(babel): version upgrade

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
Merge branch 'airman5573' of https://github.com/airman5573/react-shopping-cart-prod into airman5573

airman5573 committed:
refactor(cleanup): 사용하지 않는 module을 삭제한다

airman5573 committed:
chore(lint): lint rule을 합친다

react를 import하지 않아도 jsx를 사용할 수 있게 끔하는 rule을 extend로 합친다.

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
refactor(component): react import없이 jsx를 사용한다

airman5573 committed:
refactor(lint): lint를 적용한다

airman5573 committed:
refactor(component): react를 import하지 않고 jsx를 사용한다

airman5573 committed:
chore(lint): lint rule을 변경한다

airman5573 committed:
refactor(component): 컴포넌트를 합친다

Menu컴포넌트를 굳이 분리 할 필요가 없어서 합쳤다

airman5573 committed:
refactor(cleanup): 사용하지 않는 파일을 삭제한다

airman5573 committed:
refactor(component): 버튼 컴포넌트를 개선한다

1. type을 동적으로 입력 받을 수 있다
2. size 속성을 추가한다

airman5573 committed:
feat(component): input 컴포넌트를 추가한다

airman5573 committed:
feat(component): labeledInput 컴포넌트를 추가한다

airman5573 committed:
feat(hooks): useForm 훅을 추가한다

airman5573 committed:
feat(template): authFormTemplate을 추가한다

airman5573 committed:
style: 전역 font를 변경한다

airman5573 committed:
feat(page): 로그인 페이지를 추가한다

airman5573 committed:
feat(page): 회원가입 페이지를 추가한다

airman5573 committed:
refactor(component): 버튼에 속성을 추가한다

airman5573 committed:
refactor(component): divider에 속성을 추가한다

airman5573 committed:
refactor(component): 컴포넌트를 삭제한다

범용성 있는 FontawesomeIconButton으로 대체한다

airman5573 committed:
refactor(page): ui 개선

airman5573 committed:
fix(component): 존재하는 컴포넌트를 사용한다

airman5573 committed:
feat(page): 마이페이지를 추가한다

airman5573 committed:
refactor(hook): useForm을 개선한다

1. submit시 최근값을 돌려준다.
2.pattern validation을 지원한다

airman5573 committed:
feat: localStorage class를 구현한다

airman5573 committed:
feat(remote): remote api를 추가한다

각 페이지에서 api 요청을 보낸다

airman5573 committed:
feat(component): login 상태에 따라 header menu를 다르게 보여준다

airman5573 committed:
refactor(remote): 앱이 로드될때 로그인 상태를 업데이트한다

airman5573 committed:
feat(component): 로그인 상태에 따라 페이지 접근을 제한하는 컴포넌트를 추가한다

airman5573 committed:
refactor(cleanup): 사용하지 않는 코드를 삭제한다

airman5573 committed:
chore(lint): rule을 추가한다

alert를 허용한다

airman5573 committed:
refactor(msw): 잘못된 요청에 대응한다

airman5573 committed:
docs: 기능 요구사항 추가

airman5573 committed:
refactor: 기본값을 삭제한다

airman5573 committed:
refactor(lint): lint suggestion 해결

airman5573 committed:
refactor(cleanup): 지원하지 않는 기능을 삭제한다

airman5573 committed:
fix(component): 에러 처리를 하지 않는다

airman5573 committed:
refactor(component): button에 속성을 추가한다

disabled 속성을 추가한다

airman5573 committed:
feat(component): input 입력값에 대한 feedback을 추가한다

airman5573 committed:
fix(ui): 컴포넌트의 사이즈를 조정한다

airman5573 committed:
fix(msw): 기본 값을 넣어준다

airman5573 committed:
feat(hook): useForm hook에 custom validator 기능을 추가한다

airman5573 committed:
refactor(page): myPage의 validation을 개선한다

1. input사이의 간격을 조절한다
2. 입력값에 대한 유효성 검사 후 feedback을 보여준다

airman5573 committed:
refactor(hook): useForm에서 input의 defaultValue를 지원한다

airman5573 committed:
refactor(page): form component를 분리한다

airman5573 committed:
refactor(cleanup): 중복을 제거한다

이미 {...productObjs[id]}에서 name, thunbmail_image, price를 넘겨주고 있기 때문에
추가적으로 또 넘겨줄 필요는 없다.

airman5573 committed:
docs: 기능 구현 목록을 추가한다

airman5573 committed:
refactor: product schema를 변경한다

sku대신 id를 사용한다

airman5573 committed:
refactor(component): disabled 속성을 사용하지 않는다

airman5573 committed:
docs: 기능 구현 목록을 수정한다

airman5573 committed:
feat(msw): handler를 추가한다

장바구니, 상품 단건 조회, 주문 목록 관련 api를 추가한다

airman5573 committed:
refactor(schema): product schema를 변경한다

thumbnail_image를 thumbnailImage로 변경한다

airman5573 committed:
refactor(redux): product관련 reducer를 분리한다

airman5573 committed:
refactor(lint): 파일에서 lint rule 비활성화 한다

airman5573 committed:
refactor(redux): reducer를 분리한다

airman5573 committed:
chore(webpack): 설정을 변경한다

publicPath를 지정해서 detail 페이지에서도 root에 있는 파일을 요청하도록 한다.

airman5573 committed:
refactor(msw): error처리를 개선한다

airman5573 committed:
fix: mocking data를 정리한다

airman5573 committed:
refactor(redux): thunks를 활용한다

airman5573 committed:
refactor(naming): 함수명을 변경한다

productId가 아닌 cartItemId을 사용한다

airman5573 committed:
feat(component): 상세페이지를 추가한다

airman5573 committed:
chore(lint): rule을 추가한다

airman5573 committed:
refactor(redux): 기본 초기 상태를 추가한다

airman5573 committed:
fix(redux): cartItem의 기본 상태를 추가한다

airman5573 committed:
fix(msw): 재고 상태를 추가한다

airman5573 committed:
feat(redux): 주문 관련 reducer를 추가한다

airman5573 committed:
refactor(component): authGuard를 개선한다

로그인 상태에 따라 페이지에 접근을 제한한다

airman5573 committed:
feat(page): 주문 목록 페이지를 생성한다

airman5573 committed:
build: env파일 수정

soyi47 committed:
Merge branch 'airman5573' of https://github.com/airman5573/react-shopping-cart-prod into step2

soyi47 committed:
fix: console.log 제거

soyi47 committed:
refactor: order-list 절대경로 추가

soyi47 committed:
refactor: 사용하지 않는 파일 삭제

soyi47 committed:
refactor: AuthGuard와 사용자 접근 관련 상수 위치 이동

soyi47 committed:
refactor: useForm 위치 이동

soyi47 committed:
refactor: 사용하지 않는 파일 삭제

soyi47 committed:
refactor: LocalStorage 파일 위치 이동

- LocalStorage 클래스 정의 파일 위치를 utils로 이동
- LocalStorage import 경로 수정
- storage 절대 경로 제거

soyi47 committed:
refactor: 절대경로 shared -> components 이름 변경

- shared -> components로 절대경로 이름 변경
- 변경 파일들 import문 정리

soyi47 committed:
fix: API_URL 변경

msw URL로 변경

pocojang commented on .eslintrc.js:
```diff
@@ -29,7 +29,12 @@ module.exports = {
+    "prettier/prettier": [
+      "error",
+      {
+        endOfLine: "auto",
```
`auto`로 설정하신 이유가 있나요?!

soyi47 replied:
prettier EndOfLine default 세팅이 LF로 되어 있는데,
제 컴퓨터 윈도우에서 CRLF를 사용하고 있어서 auto로 설정해줬습니다.

pocojang commented on REQUIREMENTS.md:
```diff
@@ -1,19 +1,60 @@
+- [ ] 주문 목록 페이지
+
+  - [ ] 사용자가 주문한 주문목록을 보여준다
+
```
빈 체크는 의도되신건가요?!

soyi47 replied:
아뇨! ㅠㅠ 업데이트를 빠트렸네요. 
수정했습니다! https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/9170a63a7a686c9e9641ec7f31cd2c0fa7c6ac61

pocojang commented on src/App.jsx:
```diff
@@ -1,65 +1,66 @@
               </AuthGuard>
             }
           />
           <Route path="/product/:id" element={<ProductDetail />} />
```
- `/product`로만 접근하면 어떤 일이 발생될까요?
- `/단수/:id` vs `/복수/:id`

soyi47 replied:
1. 여러 개의 상품 중 해당 아이디를 가진 상품에 대한 접근이라는 표현을 담는 게 적절하다고 생각되어서, `복수/:id` 형태로 바꾸었습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/0f5ff1f0324ea4eab7880e1f4f984ae6b5ce8664
2. 잘못된 경로 접근에 대한 라우팅 처리를 추가했습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/9a08ed06878e7f18a995836b37f61964c0c19029

pocojang commented on src/__test__/cart.test.js:
```diff
@@ -4,27 +4,27 @@ import reducer from "../redux/reducer";
     name: "Apple Watch Series 5",
     price: 120,
-    thumbnail_image: {
+    thumbnailImage: {
```
👍

pocojang commented on src/App.jsx:
```diff
@@ -1,65 +1,66 @@
 import Login from "./pages/login/Login";
 import Signup from "./pages/signup/Signup";
 import MyPage from "./pages/my-page/MyPage";
-import AuthGuard from "./auth-guard/AuthGuard/AuthGuard";
```
뎁스가 여러번 붙는 것도 불편하지만 

```js
import ProductDetail from "./pages/product-detail/ProductDetail";
import OrderList from "./pages/order-list/OrderList";
```

이런 형태의 구조가 저는 매우 불편한다고 보는데 어떤 이유로 정의하신 컨벤션일까요?

soyi47 replied:
1. 컴포넌트 정의 파일을 폴더로 묶은 이유(ProductDetail.jsx를 product-detail 폴더로 묶어준 이유)
- ProductDetail.jsx 외에도 같은 이름의 scss, test 등의 파일이 생길 수 있어서,관련있는 파일들을 하나의 폴더로 묶어주면 더 깔끔해보인다고 생각했습니다.
- 폴더 내에 하위 컴포넌트를 정의하게 되면, 하위 컴포넌트의 네이밍을 좀 더 일반적으로 정할 수 있다는 장점이 있다고 생각했습니다.  예를 들어 product-list/Button.jsx product-detail/Button.jsx와 같이 폴더를 나눠주면 컴포넌트의 이름 충돌을 막을 수 있게 되어서 이점이 있다고 생각했어요.

2. 케밥 케이스로 폴더명을 지은 이유
- 컴포넌트가 직접 정의된 파일만 파스칼 케이스로 명시해보고자 했습니다. 그래서 컴포넌트가 아닌 (폴더)이름은 케밥 케이스로 작성해봤어요.

그런데... 막상 이 방식을 해보니 폴더 구조가 깊어지는 게 불편하게 느껴져서, 
우선 다시 플랫하게 수정해봤습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/ef5c7acbf2389882b6cbe767a194899b07c0ddbf
(옮기는 과정에서 스토리북에 에러 생겼을 것 같은데,, 곧 수정해보겠습니다.)

이후에 컴포넌트 폴더 이름을 파스칼 케이스로 하고, 내부에 index.jsx를 정의하는 방식도 생각해봤는데요.
아래와 같은 이유로 이 방식도 괜찮을 것 같다고 생각했습니다.
1. import문 뎁스가 줄어든다.
2. 컴포넌트를 정의한 파일을 찾을 때, 폴더명으로 찾으면 불편함이 없다.

컴포넌트 단위 폴더를 만드는 게 깔끔하다는 장점도 있는 것 같고,
반면 폴더 구조가 깊어지는 게 깊어지는 건 불편한 면도 있다고 느껴집니다.
어떤 구조가 좋을지 계속 고민되네요..!

pocojang replied:
여러가지 고민을 케이스로 나눠서 나열해보시니 좋습니다
그만큼 의미있는 고민을 하셨다는거 아닐까요ㅎㅎ
장단점도 있고 느끼셨으니 그것만으로도 이미 충분해요 :)

pocojang commented on src/components/auth-guard/AuthGuard.jsx:
```diff
@@ -0,0 +1,34 @@
+    }
+    if (policy === USER_ACCESS_POLICY.ONLY_LOGGED_OUT_USER && isLoggedIn) {
+      alert("접근할 수 없습니다!");
+      window.location.href = "/";
```
현재 프로젝트에서 `React Router v6`를 사용하시는데 바로 `window API`를 조작할 필요가 있을까요?

soyi47 replied:
react router를 사용하도록 수정했습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/6c2caa1ba6ba812e4c0393398654b483aab85e4f

pocojang commented on src/components/auth-guard/AuthGuard.jsx:
```diff
@@ -0,0 +1,34 @@
+}) {
+  const isLoggedIn = useSelector((state) => state.user.data.isLoggedIn);
+
+  useEffect(() => {
```
`useEffect()`가 꼭 필요하셨었나요?

soyi47 replied:
필요가 없었습니다..! useEffect 밖의 사용자 상태에 따른 분기 처리 내부에서 페이지 이동하도록 수정했습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/def02d141c3c06f2f265f528c9e64d6e673617d5

pocojang commented on src/components/number-input/NumberInput.jsx:
```diff
@@ -33,25 +24,26 @@ function NumberInput({
-    setLocalValue((prev) =>
-      positive ? Math.max(0, Number(prev) - step) : Number(prev) - step
-    );
+    const newValue = positive
+      ? Math.max(0, Number(value) - step)
+      : Number(value) - step;
+    onChange(newValue);
```
`useEffect()`가 더 어울리는 상황아닐까? 싶다라는 생각이 들기도 하네요 🧐

soyi47 replied:
1. 컴포넌트 내부에 input value 상태를 두고 제어 컴포넌트로 관리하면서 useEffect를 사용해봤었는데요. prop으로 내려 받는 onChange를 useEffect 내에서 사용하게 되었는데, 이때 문제 없이 동작하려면 onChange에 꼭 useCallback이 적용되어 있어야 했습니다. 이렇게 외부에서 컴포넌트 내부의 상황을 고려해서 로직을 작성해야 하는 상황은 피하고 싶었습니다. 
2. 또 개인적으로 useEffect가 컴포넌트 동작의 흐름을 파악하는데 불편하다고 느껴서 사용을 지양하고 싶었어요.

그래서 useEffect를 사용하지 않고 작성했습니다~

pocojang replied:
`useEffect()`를 권할만한 케이스였다는 이야기는 드리고 싶었어요 ㅎㅎ

상태를 변경하는데 어떠한 상태에 의존적인 경우 혹은 어떠한 상태의 변경을 포착하는 경우는 어떻게 동작시켜야할까?
라는거죠!

하지만!! 저 또한 `useEffect() 컴포넌트 동작의 흐름을 파악하는데 불편하다`는 @soyi47 의 의견에 동의합니다.

pocojang commented on src/components/page-title/PageTitle.stories.jsx:
```diff
@@ -16,8 +16,6 @@ export default {
 };
 
 function Template({ children, ...rest }) {
-  console.log("rest : ", rest);
-  console.log("title : ", children);
```
👍

pocojang commented on src/index.jsx:
```diff
@@ -1,13 +1,16 @@
+
+import App from "./App";
 
 function prepareMSW() {
```
해당 함수에서 검사할게 아니라...
먼저 `DEV` 환경인지 분기를 돌리고 `prepareMSW()`를 실행시키는게... 더 좋지 않을ㄲㅏ요?

soyi47 replied:
https://github.com/woowacourse/react-shopping-cart-prod/pull/40#discussion_r905187506 <- 아래 코멘트에 답 코멘트 남겼습니다!

pocojang commented on src/index.jsx:
```diff
@@ -18,40 +21,10 @@ function prepareMSW() {
-  return store;
-}
-
 prepareMSW().then(async () => {
```
좋은 아이디어이긴한데 흠 효용성이 있을까요?
`MSW`를 준비하고 실행한다는 의미를 가지는 함수가 리액트 앱 초기화와 렌더 권한을 가지고 있다는게 개인적으로는 납득하기 어려운 부분이네용

soyi47 replied:
말씀대로 prepareMSW라는 이름의 함수는, dev 환경에 대한 분기 내에서 호출되는 게 더 자연스러운 것 같아요. 또, prepareMSW의 resolve와 별개로 렌더가 이루어지는 게 맞다고 생각합니다.

유저 정보를 초기에 가져오는 로직을 Header 컴포넌트로 이동하면서, 초기 렌더와 msw 실행 로직이 따로 실행되도록 수정했습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/60b08f5e855e62a5c8d10f362f80a2a461b547c1

pocojang commented on src/index.jsx:
```diff
@@ -18,40 +21,10 @@ function prepareMSW() {
-
+  const store = createStore(reducer, applyMiddleware(thunk));
+  const accessToken = LocalStorage.getItem("accessToken");
+  accessToken && (await getUser()(store.dispatch));
```
`store.dispatch(getUser)`는 불가능한가요..?
갑자기 어렵게 스토어를 사용하는 것 같은? 코드를 보니 당황스럽긴한데 [bindActionCreators()](https://redux.js.org/api/bindactioncreators)도 사용가능합니다.

컴포넌트에서만 리덕스에 접근하다가 컴포넌트가 아닌 공간에서 호출하려니 헷갈리셨을까요?

soyi47 replied:
`store.dispatch(getUser)`가 가능했던 것으로 파악됩니다... 😂 
컴포넌트가 아닌 곳에서 리덕스를 사용하려니 어렵게 생각했던 것 같아요.

pocojang commented on src/pages/cart/components/cart-form/CartForm.jsx:
```diff
@@ -74,25 +88,21 @@ function CartForm({ className }) {
           <tbody>
-            {productIdsInCart.map((id) => {
-              const isSelected = selectedProductIds.indexOf(id) > -1;
+            {cart.map((cartItem) => {
```
`cart`에 아이템이 존재하지 않으면 어떻게 될까요?

soyi47 replied:
장바구니 아이템이 존재하지 않는 케이스에 대한 처리를 추가했습니다! https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/eef77d71906592834ab2bd6d670e3abd159af59f https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/75c328afe77bcb2755ea57885cdf8bb449344d80

pocojang commented on src/pages/cart/components/cart-form/CartForm.jsx:
```diff
@@ -74,25 +88,21 @@ function CartForm({ className }) {
-            {productIdsInCart.map((id) => {
-              const isSelected = selectedProductIds.indexOf(id) > -1;
+            {cart.map((cartItem) => {
+              const isSelected = selectedCartItemIds.indexOf(cartItem.id) > -1;
```
`indexOf()`보다 더 적절한 메서드가 있지 않을까요?

soyi47 replied:
`includes()`를 사용하도록 수정해봤습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/9a95c48245364d1a77bab4d0b7415bc950d4c353

pocojang commented on src/pages/cart/components/cart-form/CartForm.jsx:
```diff
@@ -74,25 +88,21 @@ function CartForm({ className }) {
-                      quantity={cart[id].quantity}
+                      onChecked={handleCheck(cartItem.id)}
+                      onQuantityChange={handleQuantityChange(cartItem.id)}
+                      onDelete={handleDeleteProduct(cartItem.id)}
```
`cartItem.id`에 자주 접근한다면 미리 꺼내주는게 어떨까요?

soyi47 replied:
좋습니다~ 미리 꺼내도록 수정했습니다! https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/9a95c48245364d1a77bab4d0b7415bc950d4c353

pocojang commented on src/pages/cart/components/cart-total/CartTotal.jsx:
```diff
@@ -1,21 +1,27 @@
   const handleOrderBtnClick = useCallback(() => {
-    alert("주문기능은 준비중입니다");
-  }, []);
+    dispatch(createOrder({ cartItemIds: selectedCartItemIds }));
```
상당히 명시적이네요 👍

pocojang commented on src/pages/cart/utils/getSelectedCartItemIds.js:
```diff
@@ -0,0 +1,8 @@
+function getSelectedCartItemIds(cart) {
+  return cart.reduce((acc, cartItem) => {
```
`filter()`로 가능하지 않을까요..?

soyi47 replied:
filter와 map을 사용하도록 수정해봤습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/b4f287ee1f38a81db5703b0190ec9aa3e2ae5c00

pocojang commented on src/pages/home/components/product-item/ProductItem.jsx:
```diff
@@ -1,52 +1,50 @@
   name,
   price,
-  thumbnail_image: { url: thumbnailUrl, alt },
+  thumbnailImage: { url: thumbnailUrl, alt },
```
👍

pocojang commented on src/pages/home/components/product-list/ProductList.jsx:
```diff
@@ -1,15 +1,28 @@
-      {productList.map((item) => (
-        <ProductItem key={item.sku} {...item} />
-      ))}
+      {productList &&
```
여러가지 장치를 해놓으셔서 초기값을 찾는게 쉽지 않았네요 
빈배열이 들어올때도 유효한 로직일까요?
추가적으로 `productList`의 초기 상태 혹은 어떠한 데이터도 가져오지 못했을때도 유지가 될까요?

soyi47 replied:
- 초기 상태 혹은 에러 상태에 대한 처리는 [바로 아래의 코멘트](https://github.com/woowacourse/react-shopping-cart-prod/pull/40#discussion_r904530593)에 남긴 것과 같은 방식으로, useFetch를 적용하면서 초기 상태를 바꿔줬습니다.
- 아래 코멘트의 상황과 마찬가지로 에러가 발생하지는 않을텐데, 빈 배열이 들어오는 경우에 대한 UI 처리가 없어서 추가했습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/a0f5d77ec5e69ea8bf0a39fd923b05397e3a7792

pocojang commented on src/pages/home/components/product-list/ProductList.jsx:
```diff
@@ -1,15 +1,28 @@
+  }, [dispatch]);
+
+  if (isLoading) return <div>is loading ...</div>;
+  if (isError) return <div>is error...</div>;
```
- 비슷한 로더 컴포넌트와 에러 컴포넌트가 비슷하게 계속 쓰이는데 컴포넌트화 하는게 어떨까요?
- 로직에 구멍이 없을까요? 에러도 아니고 로딩도 아니고 성공도 아니면 어떻게 될까요? 네트워크 속도를 엄청 느리게해서 한번 테스트해보세요!

soyi47 replied:
로더 컴포넌트와 에러 컴포넌트를 만들었습니다! 
1. PageLoader 컴포넌트 생성 및 적용 https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/b3e3eb67920dcad5b67b732d56b4352c2e979fd7
2. PageErrorResult 컴포넌트 생성 및 적용 https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/a969e83e74c5ca6645f3106100a748df1c32e28e

soyi47 replied:
초기 data 값으로 설정해둔 빈 배열을 불러오기 때문에, 에러가 발생하지는 않습니다. 
그래도 페이지 접속하고 바로 API 요청을 보내는 흐름이기 때문에, 지금처럼 초기에 로딩도, 에러도, 성공도 아닌 상태는 부자연스럽다고 생각이 드네요.

로딩 초기 상태를 true로 주고 시작하는 게 좋겠다고 생각했어요. 이 부분은 useFetch 커스텀 훅을 만들면서 적용시켜봤습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/38d49472c47c331985c01dee46d7389ebe00936d
(장바구니 페이지의 상태 관리는 이미 리덕스로 너무 많은 로직을 만들어둬서 걷어내기가 쉽지 않네요. 그래서,, 우선 장바구니와 유저 관련 정보는 리덕스로 관리하도록 유지했습니다. ㅠㅠ)

pocojang replied:
그렇게 확신을 가지고 대답할 수 있는 부분이면 좋습니다 :)

pocojang commented on src/pages/login/Login.jsx:
```diff
@@ -55,7 +49,13 @@ function Login({ className }) {
+              size="md"
+              block
+              type="submit"
+              disabled={isLoading}
```
에러 상황에 대한 대응은 필요없을까요?

soyi47 replied:
[userReducer의 LOGIN_REJECTED 액션](https://github.com/woowacourse/react-shopping-cart-prod/pull/40/files#diff-c5420376c331c09bb3755542979a711f5ae0744de2420e0f0e19ac5a8aa5768f)에서 에러 메세지를 alert 해주고 있습니다. 
alert 위치는 곧 옮겨보겠습니당

pocojang commented on src/pages/my-page/components/password-form/PasswordForm.jsx:
```diff
@@ -0,0 +1,84 @@
+        {...register("newPassword", {
+          pattern: {
+            value: /^.*(?=^.{8,12}$)(?=.*\d)(?=.*[a-zA-Z])(?=.*[!@#$%^&+=]).*$/,
+            message: "영문,숫자,특수문자의 조합으로 8 ~ 12글자를 입력해 주세요",
```
정규식에 익숙하지 않은 동료를 위해 조금 더 알아볼 수 있는 형태로 분리해보는게 어떨까요?
또한 직접 작성한 정규식이 아니라면 [`JSDoc See + Link`](https://jsdoc.app/tags-see.html) 조합을 사용해보세요

pocojang commented on src/pages/my-page/MyPage.jsx:
```diff
@@ -1,74 +1,15 @@
-  const dispatch = useDispatch();
-  const user = useSelector((state) => state.user, shallowEqual);
+import Divider from "@components/divider/Divider";
+import UserNameForm from "./components/username-form/UserNameForm";
```
`username-form` 뭔가 이상하지 않나요..?
네이밍 규칙을 위해 강제로 해놓으신건지요?!

soyi47 replied:
폴더 네이밍 컨벤션에 따라 케밥 케이스로 작성했는데요.

컴포넌트 폴더 하위에 바로 UserNameForm 파일이 있도록 구조를 정리하면서 해당 폴더는 지워졌습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/ef5c7acbf2389882b6cbe767a194899b07c0ddbf
케밥 케이스 네이밍 말고 다른 부분 말씀하시는 걸까요~?

pocojang replied:
딱 그 부분만인데 없어졌네요ㅎㅎ

pocojang commented on src/pages/order-list/OrderList.jsx:
```diff
@@ -0,0 +1,48 @@
+    orderList: state.orderList.data,
+  }));
+
+  console.log("orderList in OrderList.jsx : ", orderList);
```
😢

soyi47 replied:
console.log 제거했습니다. 😢 https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/304383c73a9018f3647b4e476e2d60f69b6969b4

pocojang commented on src/pages/order-list/components/order-table-item/order-table-item.module.scss:
```diff
@@ -0,0 +1,23 @@
+    }
+    .right {}
+  }
+}
```
😢

soyi47 replied:
사용하지 않는 클래스는 삭제해주었습니다~ 😥 https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/76efad33c68b7393af0e074e40c26b2dc7d5c5b0

pocojang commented on src/pages/order-list/components/order-table/order-table.module.scss:
```diff
@@ -0,0 +1,15 @@
+
+table {
+  width: 100%;
+}
```
TODO ?!

soyi47 replied:
TODO였으나,,, 작성하지 않을 것 같아 삭제해주었습니다~ 😥 https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/76efad33c68b7393af0e074e40c26b2dc7d5c5b0

pocojang commented on src/pages/product-detail/ProductDetail.jsx:
```diff
@@ -1,9 +1,57 @@
+    dispatch(getProductDetail({ productId }));
+  }, [dispatch, productId]);
+
+  if (isLoading) return <div>loading...</div>;
```
에러 케이스는 다루지 않으시는걸까요?

soyi47 replied:
에러 케이스에서는 PageErrorResult 컴포넌트를 렌더하도록 처리해줬습니다. https://github.com/woowacourse/react-shopping-cart-prod/commit/a969e83e74c5ca6645f3106100a748df1c32e28e

pocojang commented on src/pages/product-detail/product-detail.module.scss:
```diff
@@ -0,0 +1,32 @@
+
+.productPrice {
+  font-size: 1.25rem;
+}
```
`scss`는 포맷팅이 안되고 있으신가요..?

soyi47 replied:
몰랐는데 그랬네요 😥 포맷팅 적용해줬습니다..! https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/05174f6abba66fbaccf41abad47d843742011a91

pocojang commented on src/redux/constants.js:
```diff
@@ -0,0 +1,59 @@
+import queryState from "./utils/queryState";
+
+export const initialState = {
```
아이디어는 좋은데 이 녀석의 정체가 궁금하네요
복잡도가 굉장히 늘어납니다

soyi47 replied:
A 리듀서에서 관리하는 상태를 B 리듀서에서 확인해야하는 상황이 있어서, [전역 상태 전체의 초기 상태](https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/b60e76daebb0889e6a1db6599c8bdfd8833cb185#diff-65f4acba9334b506fe87a1567271a29d61f39cc25762134b75489b23e03ef266)를 만들고, [combineReducer 없이 리듀서를 구분해주는 상위 리듀서를 작성](https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/b60e76daebb0889e6a1db6599c8bdfd8833cb185#diff-303de201ad48140bd3a4da20317ae183d128c0fbd0edf31335deb608e135991e)했습니다. 리듀서를 구분할 때 전체 전역 상태를 totalState라는 인자로 내려 받아 확인할 수 있도록 했었고요..!

전역 상태를 줄이면서 하나의 리듀서에서 다른 리듀서의 상태를 확인할 필요는 없어져서, combineReducer를 사용하고 상태를 리듀서 단위로 나누어 관리하도록 정리했습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/07a9488894750906151b65e8bfcd3b65e834881d

pocojang commented on src/redux/constants.js:
```diff
@@ -0,0 +1,59 @@
+      price: 0,
+      thumbnailImage: {
+        url: "https://via.placeholder.com/300",
+        alt: "Empty Image",
```
`Empty Image`가 맥락상 맞을까요?

soyi47 replied:
`Placeholder Image`가 더 적절했을 것 같네요!
(해당 코드는 productDetail 관련 전역 상태를 걷어내면서 같이 없어졌습니다)

pocojang commented on src/redux/constants.js:
```diff
@@ -0,0 +1,59 @@
+      secession: queryState.init(),
+      getUser: queryState.init(),
+      updateUserPassword: queryState.init(),
+      updateUserGeneralInfo: queryState.init(),
```
`query`라고 하기에는 질의하는게 아예 없지 않나용?

soyi47 replied:
query라는 단어에 대해서 고민을 해봤는데, 
데이터베이스에 대한 정보 요청이라는 의미의 query 보다는 API 서버로의 요청이라는 의미가 더 적절한 것 같네요. 
queryState -> apiRequestState로 이름을 바꿔줬습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/4c71a7468bd535cda02fef2a0b5cb9446970fb7a

다른 부분에 대한 질문이셨다면 말씀 부탁드려요!

pocojang commented on src/redux/reducer.js:
```diff
@@ -1,92 +1,66 @@
-import ACTION_TYPE from "./actions";
+/* eslint-disable default-param-last */
```
어떤 이유로 사용하셨나요?

soyi47 replied:
`default-param-last`이 마지막에 오는 인자에만 기본값을 줄 수 있도록 하는 설정인데요. 
리듀서에서는 state, action 순으로 인자를 받는데, 첫번째로 오는 인자인 state의 기본값으로 initialState를 줘야 하는 상황이라, 해당 파일에서만 `default-param-last` 설정을 disable 시켰습니다.

pocojang commented on src/redux/reducers/cart-reducer/cartReducer.js:
```diff
@@ -0,0 +1,142 @@
+    case ACTION_TYPE.GET_CART_REJECTED: {
+      const newState = structuredClone(state);
+      const { error } = payload;
+      alert(error.message);
```
어울리는 위치가 맞을까요?

soyi47 replied:
성공 또는 실패의 상태가 확정되었을 때, 리덕스 스토어에 대한 처리와 동시에 알림을 띄우려고, 리듀서에 alert를 넣게 되었는데요.

리듀서를 액션에 대한 처리 로직이라는 관점에서 보면 위의 방식이 괜찮다는 생각이 들었고, 반대로 전역 상태 관리를 위한 로직으로 보면 상태 관련 로직만 들고 있는 게 괜찮을 것 같아 계속 고민이 되었습니다...

성공 또는 실패 상태를 내려 받아 컴포넌트 단에서 alert하는 것이 더 적절하다고 보시는 걸까요? 
혹은 thunk에서 성공 또는 실패 액션을 dispatch하면서 alert 하는 건 어떻게 생각하시나요?

pocojang replied:
`Redux`의 기본 원칙 그리고 `Reducer`는 정확히 어떤 역할만 해야할까? 고민해보시고 잘 지키고 있나 생각해보시면 좋지요

저는 말씀하신 것 중 후자에 가깝게 생각합니다ㅎㅎ
>성공 또는 실패 상태를 내려 받아 컴포넌트 단에서 alert하는 것이 더 적절하다고 보시는 걸까요?
혹은 thunk에서 성공 또는 실패 액션을 dispatch하면서 alert 하는 건 어떻게 생각하시나요?

pocojang commented on src/redux/reducers/cart-reducer/cartReducer.js:
```diff
@@ -0,0 +1,142 @@
+import queryState from "@redux/utils/queryState";
+import ACTION_TYPE from "./cartActions";
+
+const cartReducer = (state, { type, payload }, totalState) => {
```
개인적으로는 `cartReducer`가 하는 일이 너무나 많다는 생각이 듭니다.

soyi47 replied:
1. 액션이 너무 많다.
2. 각 액션에 대한 로직 내부에서 데이터를 조작하는 처리가 너무 많다.

두 가지 의미로 말씀하신 것으로 파악했는데요!

1. 액션이 많아진 부분은 API 요청에 대한 처리 그리고 그로 관리하는 장바구니 아이템 상태에 대한 처리를 모두 리덕스를 거치기로 한 이상 어쩔 수 없었던 것 같아요. ㅠㅠ 혹은 리듀서 단위를 더 쪼개는 게 나았겠다는 말씀일까요?
2. 지금은 데이터 조작 로직이 thunk와 reducer에 분산되어 있다는 느낌이 듭니다... thunk를 사용하는 경우라면, thunk에서 데이터를 조작하고 조작한 값을 액션을 dispatch 하면서 넘겨주어 바로 상태를 업데이트 하는 게 더 깔끔했을 것 같습니다.

pocojang commented on src/redux/reducers/cart-reducer/cartThunks.js:
```diff
@@ -0,0 +1,147 @@
+export const defaultDeleteCartItemsThunkErrorMessages = {
+  3002: "토큰이 만료되었거나 존재하지 않습니다.",
+  6002: "장바구니에 존재하지 않는 아이템입니다.",
+};
```
정리 좋습니다👍 
다만 이런 코드가 여기있어야하는지는 의문이네요

soyi47 replied:
에러 메시지 상수 파일을 분리해줬습니다~ https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/5405462e83aa904640521e6c616ad18782605198

pocojang commented on src/redux/reducers/order-reducer/orderReducer.js:
```diff
@@ -0,0 +1,33 @@
+      const newState = structuredClone(state);
+      newState.query.createOrder = queryState.fullfilled();
+      alert("주문이 성공적으로 들어갔습니다.");
+      window.location.href = "/order-list";
```
음.. 리듀서에 있어야하는 코드가 맞을까요?

soyi47 replied:
https://github.com/woowacourse/react-shopping-cart-prod/pull/40#discussion_r904603421 <- 같은 상황에 대한 코멘트가 이미 있어서, 해당 코멘트에 질문 남겨뒀습니다!

pocojang commented on src/utils/fetcher.js:
```diff
@@ -0,0 +1,65 @@
+import LocalStorage from "@utils/LocalStorage";
+
+class Fetcher {
```
온통 `Static`메서드인데 `Class`인 이유가 있을까요?

soyi47 replied:
특별한 이유는 없고,,, Fetcher에서 accessToken을 멤버 변수로 들고 관리하도록 해보려고 class를 사용해봤습니다.

pocojang replied:
구현된 코드를 봤을때는 `Class`로 작성될 이유를 찾기가 어려웠네요 ㅎㅎ

pocojang commented on src/hooks/useForm.jsx:
```diff
@@ -1,4 +1,4 @@
-import { useState, useRef } from "react";
+import { useState, useRef, useEffect } from "react";
 
 function useForm() {
```
그 유명한 `react-hooks-병민-form`을 또 만났군요..

pocojang commented on src/hooks/useForm.jsx:
```diff
@@ -24,6 +24,14 @@ function useForm() {
           }
         }
+        if (validation.customValidator) {
+          const { isValid, errorMessage } = validation.customValidator(
```
이런 시도는 좋읍니다 👍

pocojang commented on src/hooks/useForm.jsx:
```diff
@@ -61,27 +82,41 @@ function useForm() {
           ref,
           validation: {
             pattern: options?.pattern,
+            customValidator: options?.customValidator,
```
기본 값 처리는 안해도 될까요?

pocojang reviewed (changes requested):
@soyi47 👋 
대타로 왔읍니다!

## 사용성

당연히 잔버그가 있습니다!
일정지키느라 힘드셨죠? 직접 사용해보며 찾아보고 수정해보세요!

## Hooks

사용률이 거의 제로에 가깝네요ㅠㅠ
아쉽습니다! 더 많이 삽질하고 느껴보셨으면 좋겠습니다
다음 피드백에서는 상당 부분 커스텀 훅으로 리팩터링해보세요 :)

## 라이브러리 잘 활용하기

라이브러리를 사용한다면 그 라이브러리를 잘 활용하시는 것이 팀 컨벤션에도 좋다고 생각합니다.
라이브러리의 온전한 기능을 사용하지 않는 경우가..

`Redux`, `Router` 등이 그 예인데요.
기본적인 기능을 제공해주는데 왜 직접 구현했지??
하는 부분이 계속 보입니다.

## Redux

위의 API 활용하기와 겹치는 내용이 꽤 있을 것 같고요.
`Redux`의 기본 원칙을 어기는 부분이 반복됩니다.

그러다보니 `Reducer`가 굉장히 비대하고 `Action`에서 하는 일도 굉장히 많습니다.
`Thunk`가 주는 편리함이 오히려 부메랑으로 다가오는 듯 하고요.

결국 전체적인 비즈니스 로직이 `Redux`에 오로지 집중되어 있는 형태입니다.

물론 이 피드백은 `Redux`가 주류를 이루던 시대라면 할 수 없었겠지만 그동안 `React`가 상당 부분 발전하고 `Hooks API`가 태어나면서 `Redux`가 없이도 다른 `React`의 기능들만으로 아이디어를 낼 수 있는 부분들이 많은데 대부분 활용되지 않은 것 같아서 아쉬움이 듭니다.

## 컴포넌트 초기화 및 데이터 페치 상태 관리

데이터 초기값 그리고 로딩, 성공, 실패에 구멍이 종종 보입니다.
사용자 입장에서 다양한 경우의 수로 컴포넌트를 동작시켜보세요!

## 복잡도

좋게 말하면 코드 장인처럼 다양한 시도를 하셨는데 나쁘게 말하면..
코드 읽기가 굉장히 힘들었습니다.

불필요한 추상화 불필요한 계층이 반복적으로 나타나서 코드 읽고 이해하는데 한참 걸렸습니다.
코드를 타고 또 타고 타고 또 타고 찾아야하는거죠.

코드를 작성하는 개발자 입장에서 봤을때는 많이 배우고 이것저것 시도하기에 좋았을 것 같지만 함께 일하는 동료 입장에서 봤을때는.. 디버깅도 어렵고 컨벤션 예측이나 코드의 흐름 파악에 어려움이 있습니다.

---

개인적으로 4기에서 진행했던 코드 리뷰 중 가장 오래 걸렸습니다...
제가 설계한 미션이고 병민 코드 리뷰에서 봤던 내용들이 보여서 빠르게 끝낼 수 있을 것 같았는데... 와 정말 쉽지 않았습니다.
결국 막차타고 집에 와서 또 리뷰 정리합니다.

마지막 미션이라 전체적인 코드 리뷰에 개인적인 피드백과 견해를 마구마구 담아냈으니 자유롭게 피드백 반영해보세요.
화이팅 💪

soyi47 committed:
refactor: 파일 구조 및 파일 이름 변경

- 폴더 중첩을 줄이기 위해 컴포넌트 단위 폴더 최대한 삭제
- scss 파일 이름을 컴포넌트 jsx 파일과 같이 파스칼 케이스로 변경
- stories 폴더 생성 및 stories 파일 이동
- 페이지 단위 절대 경로 삭제

soyi47 committed:
fix: console.log 삭제

soyi47 committed:
refactor: 페이지 내 컴포넌트 단위 폴더 제거 및 파일 이동

soyi47 committed:
refactor: getSelectedCartItemIds 내부 로직 변경

reduce 대신 filter와 map 사용하도록 변경

soyi47 committed:
refactor: CartForm 장바구니에 담긴 상품이 없는 경우 UI 추가

soyi47 committed:
refactor: CartForm 장바구니 상품 목록 생성 로직 수정

- cartItem 구조 분해 할당
- isSelected 값을 includes 메서드를 사용하여 만들도록 변경

soyi47 committed:
chore: package.json 내부 url 변경

soyi47 committed:
refactor: 페이지 단위 컴포넌트 이름 변경

Page postfix 추가

soyi47 committed:
refactor: ProductList 컴포넌트 삭제

ProductList 컴포넌트의 로직을 ProductListPage로 이동

soyi47 committed:
refactor: 일부 전역 상태 제거 및 useFetch 추가

- productList, productDetail, orderList 전역 상태 제거
- 전역 상태에서 제거한 비동기 처리는 useFetch를 사용하도록 수정
- 컴포넌트 CartForm -> CartItemList로 이름 변경

soyi47 committed:
refactor: PageLoader 컴포넌트 생성 및 로딩 상태의 페이지 렌더링에 적용

soyi47 committed:
refactor: PageErrorResult 생성 및 에러 상태의 페이지 렌더링에 적용

soyi47 committed:
refactor: 유저 정보 초기 세팅 위치를 Header 컴포넌트로 이동

- index.js의 msw 실행 설정 코드 수정
- accessToken을 이용한 유저 정보 초기 설정을 Header 컴포넌트에서 하도록 변경
- user 전역 상태에 LOGOUT 액션 추가

soyi47 committed:
refactor: input 상태 관리를 위한useForm 제거 및  useInput 추가

- useForm을 useInput으로 대체
- 입력값 별 validator 함수 작성

soyi47 committed:
fix: form 내부 컴포넌트 간격 조절

soyi47 committed:
refactor: 컴포넌트 이름 변경

- CartForm -> CartItemList
- SecessionForm -> WithdrawalForm

soyi47 committed:
fix: 오타 수정

- fullfilled -> fulfilled

soyi47 committed:
refactor: AuthGuard 페이지 이동 로직 수정

- AuthGuard 컴포넌트에서 페이지 이동 시 react-router의 useNavigate 사용하도록 변경

soyi47 committed:
fix: 로그인 여부에 따라 노출되는 메뉴 변경

- 장바구니 메뉴는 로그인 시에 보이도록 수정
- 주문목록 메뉴 추가

soyi47 committed:
refactor: useFetch의 response 상태 추가

data 상태와 response 상태를 따로 가지고 있도록 하여, 데이터 값이 없는 응답에도 대응할 수 있도록 수정

soyi47 committed:
refactor: orderReducer 제거

- CartTotal 컴포넌트에서 orderReducer 대신 useFetch를 사용하도록 수정

soyi47 committed:
refactor: useFetch param 변경 및 내부 fetch 함수 param 추가

- requestData를 fetch 함수의 param으로 변경
- useFetch param을 객체 형태로 변경

soyi47 committed:
fix: mock handler 오류 수정

필요없는 JSON.parse 제거

soyi47 committed:
refactor: 사용하지 않는 param, 초기 상태 제거

- redux 전역 상태 중 order 관련 삭제
- cartReducer에 param 정리

soyi47 committed:
docs: REQUIREMENTS.md 업데이트

soyi47 committed:
refactor: 사용하지 않는 스타일 클래스 코드 삭제

soyi47 committed:
refactor: queryState -> apiRequestState로 변경

soyi47 committed:
refactor: 사용하지 않는 scss 파일 삭제

soyi47 committed:
fix: scss 포맷팅

soyi47 committed:
refactor: 상품 상세 페이지 링크 변경

- product/:id -> products/:id 복수형으로 변경

soyi47 committed:
fix: 잘못된 경로 접근에 대한 처리 추가

- WrongAccessPage 컴포넌트 생성
- 지정되지 않은 경로 접근의 경우 WrongAccessPage로 이동하도록 라우팅 처리

soyi47 committed:
refactor: AuthGuard 내부 useEffect 제거

useEffect 없이 사용자 로그인 상태 분기 처리 내에서 페이지 이동하도록 수정

soyi47 committed:
refactor: USER_ACCESS_POLICY 내부 속성 이름 변경

이미 상위 객체 이름에 USER가 있으으로, 속성 이름 마지막에 붙은 USER 삭제
ONLY_LOGGED_IN_USER -> ONLY_LOGGED_IN
ONLY_LOGGED_OUT_USER -> ONLY_LOGGED_OUT

soyi47 committed:
refactor: CartItemList 장바구니 아이템이 없는 케이스 처리 로직 수정

soyi47 committed:
refactor: 상품 목록에 상품이 없는 케이스에 대한 UI 추가

soyi47 committed:
refactor: reducer 폴더 이름 변경 및 combineReducer 적용

soyi47 committed:
refactor: ApiError 폴더 이동

soyi47 committed:
refactor: useFetch에서 ApiError 던지도록 수정

응답 데이터가 있어야 하는데 없는 경우, ApiError를 던지도록 수정

soyi47 committed:
fix: 상품 상세 페이지 장바구니 담기 기능 오류 수정

soyi47 committed:
refactor: HeaderMenus 컴포넌트 분리

soyi47 committed:
refactor: LabeledInput 내부 Input 컴포넌트 위치 이동

soyi47 committed:
refactor: Header에 menu props 추가

menu prop으로 HeaderMenus 받아 사용하도록 수정

soyi47 committed:
refactor: AuthGuard 컴포넌트 로직을 useAuthGuard 커스텀 훅으로 변경

페이지 단위에서 useAuthGuard 커스텀 훅을 사용하도록 변경하여, App.jsx로의 의존성을 제거

soyi47 committed:
fix: 오타 수정

soyi47 committed:
refactor: 에러 메시지 상수 파일 분리

soyi47 committed:
refactor: cartReducer 내부 alert 위치 변경

alert 문을 thunk로 모두 이동

soyi47 committed:
refactor: userReducer 내부 alert 위치 변경

alert문과 window.location 문을 thunk로 모두 이동

soyi47 commented:
안녕하세요~ @pocojang, @Vallista 리뷰어님들~ 소피아입니다. 🥰
방학이라 다른 일정들도 많아서 리뷰 반영이 꽤 늦어졌습니다.
리뷰 받은 내용 대부분을 반영했으니 여유 되실 때 확인 부탁드려요. 👍

두 분 다 제가 더 고민하고 생각을 정리해야 할 내용들을 짚어주셔서 좋았어요! 감사합니다. 🙇‍♀️

#### @pocojang
- 안녕하세요. 포코! 사실 포코도 리뷰 달아주실 거라 생각 못 했었어요... 하핳 빠르고 꼼꼼한 리뷰 감사드립니다. 😆
새롭게 작성해 본 스타일과 시도는 제게 의미가 있었다고 생각이 들지만, 규모가 커지니 분명 코드를 짰던 저도 헷갈리는 부분들이 생기네요. 앞으로 기본 원칙과 일관성, 다른 사람이 보기에도 코드 흐름 파악이 쉬운가를 더욱 고려하며 작성해 보겠습니다!
- 대부분의 코멘트 아래에 수정 사항 또는 질문 코멘트 남겨뒀습니다.
- 리덕스의 기본 원칙과 리듀서의 역할을 실제 코드로 어떻게 적용하는 것이 옳은가에 대해서는 제 생각이 정립이 덜 된 것 같아요. 남은 방학 동안 더 생각해 보겠습니다. 😥 우선 최대한 상태 수정 관련 로직만 리듀서에서 들도록 하고 싶어서, alert나 window.location 관련 코드는 thunk로 옮겨줬습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/b04405831f7f7b617cdd5dbf6dc495bacf73c864 https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/3ac9b46f38c76df36296bd4634ae16dba16c2e87

#### @Vallista
- 안녕하세요. 발리스타~! 월요일 오프라인 리뷰 감사했습니다. 😄 
이야기를 재미있고 명쾌하게 해주셔서 리뷰 시간이 즐거웠어요! 라이브러리에 대한 의존과 상태 관리, 관심사의 분리에 대해 이야기 나누면서, 지금까지 리액트로 작성한 코드 전반을 돌아보며 고민해 볼 수 있어 좋았습니다. 앞으로도 그날 해주신 이야기들 계속 떠올려보며, 어떻게 코드에 녹일 수 있을지 고민해 볼게요!
리뷰 시간에 제 생각을 빠르게 정리하고 여러 질문을 하면 좋았을 텐데, 느린 두뇌 회전으로 그러지 못했던 것 같아 스스로에게 아쉬움이 있어요. ㅎㅅㅎ 다음에 기회가 된다면 더 적극적인 리뷰이의 자세를 가져보겠습니다.
- AuthGuard 컴포넌트의 로직을 useAuthGuard 커스텀 훅으로 옮기고, 페이지 단위 컴포넌트에서 해당 커스텀 훅을 사용하도록 변경하여 App.jsx에서의 의존성을 줄였습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/60429df94f6042ff58584b6b9eedf2bc45192915
- Label로 Input을 감싸는 방식으로 수정하여, id 속성을 주지 않고 Label과 Input을 연결하도록 수정했습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/ef3817d183d793afbea3786d6a47abe7c489a6ce
- Header 컴포넌트에서 메뉴 영역에 해당하는 컴포넌트 HeaderMenus를 분리했습니다. 인증, 사용자 관련 로직은 HeaderMenus 컴포넌트에서만 들고 있도록 하였고요. menu라는 prop으로 해당 컴포넌트를 받아 메뉴 영역에 렌더링 하도록 수정했습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/1589b033b59df7d6d5a6d4a62494e2035ffed071 https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/3212aa5b2b9c1a7a8854f7509ceeebd2532156c3
- useFetch 내부에서 기존에 있던 커스텀 에러 ApiError를 사용하도록 수정했습니다. 에러 코드와 에러 메시지를 받는 커스텀 에러인데, 에러 코드와 메시지는 다른 에러 코드나 메시지와 겹치지 않도록 임의로 지정했습니다. https://github.com/woowacourse/react-shopping-cart-prod/pull/40/commits/18a1e964effc913dae4784738a5d557f230f6f89

pocojang reviewed (approved):
@soyi47 
광휘님에게 리뷰받았으면 이미 충분합니다ㅎㅎ
제 리뷰는 일종의 소방수였을뿐이니 가볍게 무시하시고요

일정상 이번 PR은 이만 머지하겠습니다 💯 

그동안 정말 어렵고 힘드셨죠?
레벨2...많이 삽질하고 많이 고민하셨다면 레벨3~4에서 더 좋은 코드를 작성하실 수 있을겁니다.
방학이 벌써 얼마안남았네요 😢
남은 기간만이라도 푹 쉬시길.. 🍀
